handshake: test performance with OSSL_LIB_CTX per thread

npajkovsky opened the pull request:
```
$ id
uid=0(root) gid=0(root) groups=0(root)

$ ./handshake -p /Users/npajkovsky/openssl/openssl/test/certs 500
Average time per handshake: 23085.715500us
Handshakes per second: 5651.591799
```

Resolves: https://github.com/openssl/project/issues/1200

Sashan commented on source/perflib/perflib.h:
```diff
@@ -41,6 +41,10 @@ int perflib_create_ssl_ctx_pair(const SSL_METHOD *sm, const SSL_METHOD *cm,
                                 int min_proto_version, int max_proto_version,
                                 SSL_CTX **sctx, SSL_CTX **cctx, char *certfile,
                                 char *privkeyfile);
+int perflib_create_ossl_lib_ctx_pair(OSSL_LIB_CTX *libctx, const SSL_METHOD *sm,
```
What is the reason to introduce `perflib_create_ossl_lib_ctx_pair()`? I think the `libctx` argument can be just added to existing function `perflib_create_ssl_objects()`. Caller can then pass either NULL or pointer to libctx.

the change looks good otherwise. this is the only question I have.

thanks.

Sashan reviewed (approved):
looks good to me.

I don't feel strongly about adding new function or modifying the existing one. either way is fine with me.

FYI: I've hit the thread key local storage limitation:
```
lifty$ ./handshake -t -p /home/sashan/work.openssl/QUIC/openssl.addr_validation/test/certs 100
/home/sashan/work.openssl/perftools/source/handshake.c:100: Failed to create ossl lib context
/home/sashan/work.openssl/perftools/source/handshake.c:100: Failed to create ossl lib context
/home/sashan/work.openssl/perftools/source/handshake.c:100: Failed to create ossl lib context
/home/sashan/work.openssl/perftools/source/handshake.c:100: Failed to create ossl lib context
/home/sashan/work.openssl/perftools/source/handshake.c:100: Failed to create ossl lib context
/home/sashan/work.openssl/perftools/source/handshake.c:100: Failed to create ossl lib context
```
My system is OpenBSD 7.6

npajkovsky commented:
@Sashan I have updated the README file with a note about the test limitations, and added doc for a new `-p` option.

There was no objection about the not merge it.

Sashan commented on README:
```diff
@@ -55,13 +55,17 @@ and two required arguments:
 and the average number of simultaneous handshakes per second performed over the
 course of the test.
 
+Note: ossl_lib_ctx per thread with higher number of threads will hit the thread
```
I don't know how others feel about it. In my opinion we should not be adding such note here. I think the thread local storage issue should get fixed rather quickly.

depending on whether the fix to libcrypto will be backported to release branches we might want to have such note saying something like: `Note on OpenSSL earlier than ?3.6? you might hit issue ....`

npajkovsky replied:
Ok, I'll drop it.

t8m commented on source/handshake.c:
```diff
@@ -140,17 +196,27 @@ int main(int argc, char * const argv[])
-            printf("Failed to create SSL_CTX pair\n");
+    if (per_thread_ossl_lib_ctx) {
+        int ret = perflib_run_multi_thread_test(do_handshake_ossl_lib_ctx_per_thread, threadcount, &duration);
+        if (!ret) {
```
Nit: Add empty line after declarations and split long lines.

npajkovsky replied:
Thanks, updated.

t8m replied:
You did not add the empty line after the declaration.

npajkovsky replied:
I have updated the MR.

t8m reviewed (changes requested):
One nit, otherwise LGTM

t8m commented:
Merged. Thank you.
